haystack/core/component/component.py

```diff
@@ -160,12 +160,29 @@ def run(self, **kwargs):
         isinstance(MyComponent, Component)
     """
 
-    # This is the most reliable way to define the protocol for the `run` method.
-    # Defining a method doesn't work as different Components will have different
-    # arguments. Even defining here a method with `**kwargs` doesn't work as the
-    # expected signature must be identical.
-    # This makes most Language Servers and type checkers happy and shows less errors.
-    run: Callable[..., Dict[str, Any]]
+    # The following expression defines a run method compatible with any input signature.
+    # Its type is equivalent to Callable[..., Dict[str, Any]].
+    # See https://typing.python.org/en/latest/spec/callables.html#meaning-of-in-callable.
+    #
+    # Using `run: Callable[..., Dict[str, Any]]` directly leads to type errors: the protocol would expect a settable
+    # attribute `run`, while the actual implementation is a read-only method.
```

Comment:
This can be addressed by defining `run` as a property getter instead: https://github.com/deepset-ai/haystack/pull/9344

Reply:
Yes, yours also seems like a valid solution.
The one proposed in this PR seems more explicit to me, but I would say it's a matter of taste.